### PR TITLE
new resource: `azurerm_private_dns_resolver `

### PR DESCRIPTION
**Pull request by ms-henglu**
https://github.com/hashicorp/terraform-provider-azurerm/issues/16956

### CODE REVIEW

**Review by katbyte — CHANGES_REQUESTED** [review]
you may need to configure what regions to run this service in:
```
Test Failed

------- Stdout: -------
=== RUN   TestAccDNSResolverDnsResolver_requiresImport
=== PAUSE TestAccDNSResolverDnsResolver_requiresImport
=== CONT  TestAccDNSResolverDnsResolver_requiresImport
=== CONT  TestAccDNSResolverDnsResolver_requiresImport
    testcase.go:110: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating Dns Resolver (Subscription: "*******"
        Resource Group Name: "acctest-rg-220922165907733233"
        Dns Resolver Name: "acctest-dr-220922165907733233"): performing CreateOrUpdate: dnsresolvers.DnsResolversClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="BadRequest" Message="The provided location 'westus2' is not available for Microsoft.Network/dnsResolvers and Microsoft.Network/dnsForwardingRulesets resources." Target="" Details=[{}] InnerError={}
        
          with azurerm_dns_resolver.test,
          on terraform_plugin_test.tf line 20, in resource "azurerm_dns_resolver" "test":
          20: resource "azurerm_dns_resolver" "test" {
        
        creating Dns Resolver (Subscription: "*******"
        Resource Group Name: "acctest-rg-220922165907733233"
        Dns Resolver Name: "acctest-dr-220922165907733233"): performing
        CreateOrUpdate: dnsresolvers.DnsResolversClient#CreateOrUpdate: Failure
        sending request: StatusCode=0 -- Original Error: Code="BadRequest"
        Message="The provided location 'westus2' is not available for
        Microsoft.Network/dnsResolvers and Microsoft.Network/dnsForwardingRulesets
        resources." Target="" Details=[{}] InnerError={}
--- FAIL: TestAccDNSResolverDnsResolver_requiresImport (70.39s)
```

**Comment by ms-henglu** [comment]
Hi @katbyte , maybe disable the `ARM_PROVIDER_DYNAMIC_TEST `?

**Comment by katbyte** [comment]
No, we can configure this service to only run in certain regions grep for:
```
// Server is only available in certain locations
        "analysisservices" to testConfiguration(locationOverride = LocationConfiguration("westus", "northeurope", "southcentralus", true)),
```

**Comment by ms-henglu** [comment]
Hi @katbyte , thanks for the suggestion, I've added it into test configs. Please take another look, thanks!

**Comment by sebader** [comment]
The service itself is very specially called "Azure DNS Private Resolver" (not just DNS resolver). I'd highly recommend to align the name of the terraform resource, too. Or is there a reason to only call it azurerm_dns_resolver?

https://learn.microsoft.com/en-us/azure/dns/dns-private-resolver-overview

**Comment by ms-henglu** [comment]
Hi @sebader , thanks for the suggestion! I've confirmed with the service's PM, he prefers `azurerm_dns_resolver`, which aligns with azure cli and powershell.

**Comment by tombuildsstuff** [comment]
@ms-henglu are there plans for this service to also support the Public DNS resources? If not this'd make more sense as `Private DNS` since it's scoped at that?

**Comment by katbyte** [comment]
@ms-henglu - i agree with @sebader, we should match the correct name and change this to azurerm_dns_private_resolver

**Comment by sfiguemsft** [comment]
@katbyte @sebader we're looking for consistency with naming which is already in place for PSH and CLI clients

**Comment by ms-henglu** [comment]
Hi @katbyte , WDYT about 
> we're looking for consistency with naming which is already in place for PSH and CLI clients

**Comment by katbyte** [comment]
@ms-henglu - are there plans to support public DNS with this service? because if not this the name is misleading and conflicts with the existing nomenclature in the provider with azurerm_private_dns_* for private DNS, and azurerm_dns for public DNS

**Comment by sfiguemsft** [comment]
> @ms-henglu - are there plans to support public DNS with this service? because if not this the name is misleading and conflicts with the existing nomenclature in the provider with azurerm_private_dns_* for private DNS, and azurerm_dns for public DNS

@katbyte yes, public version will still be under this same service

**Review by katbyte — APPROVED** [review]
Thanks both - with public going into the same service this LGTM now 🙌

**Comment by desweil** [comment]
@JannoTjarks It is probably true that there never will be a "public resolver SAAS Services". Still I think the name is fine, because it is "THE" Azure DNS Resolver and it already is named like this in PSH (see for [example](https://learn.microsoft.com/en-us/powershell/module/az.dnsresolver/new-azdnsresolver?view=azps-9.0.1) and AZ CLI (see for [example](https://learn.microsoft.com/en-us/cli/azure/dns-resolver?view=azure-cli-latest) from MS.

**Comment by sfiguemsft** [comment]
> > Thanks both - with public going into the same service this LGTM now 🙌
> 
> Hi @katbyte, I think there is a misunderstanding for this resource. I think, there will never be something like a azure public dns resolver, because there is no need. Microsoft has two possible dns zones: Public DNS und Private DNS. Public DNS can be resolved from the internet. Private DNS Zones can only be resolved from a virtual network in your specific azure tenant. So if you want to resolve a private dns record from your on-premise network, you need a dns resolver in azure for the private dns records. The Azure Private DNS Resolver is Microsoft SaaS Solution for this.
> 
> As I mentioned, public dns can resolved from everyone from anywhere. There is no SaaS from Microsoft needed.
> 
> So I would recommend to use the correct name azure dns private resolver.

@JannoTjarks as product owner of the services mentioned above, I prefer not to disclose any information on this PR related to roadmap and/or directional strategy of Azure DNS. Lets keep consistency with CLI/PSH were possible given that this is the only DNS resolver service from Azure

**Comment by JannoTjarks** [comment]
Hi @sfiguemsft thanks for the clarification 👍🏻 I understand that the product owner has more insights :D

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Based on responses here and how the naming is in the portal/docs, the name should not be `azurerm_dns_resolver` and instead be `azurerm_private_dns_resolver`. We should keep it consistent with the other dns resources in Terraform and not psh and cli.

**Comment by sfiguemsft** [comment]
> Based on responses here and how the naming is in the portal/docs, the name should not be `azurerm_dns_resolver` and instead be `azurerm_dns_private_resolver`. We should keep it consistent with the other dns resources in Terraform and not psh and cli.

@mbfrahry I'm asking the name to be azurerm_dns_resolver and this is what I would like customers to see in the TF modules - if this needs to be formalized through other channel please let me know.

**Review by katbyte — CHANGES_REQUESTED** [review]
@sfiguemsft - given this is for private DNS only i also agree this should be `azurerm_private_dns_resolver` - our resources names often do not 1:1 match service names and generally are chosen to create the least confusion for our users. Primarily we look to the portal for guidance on what to pick as that is where most people will be using the service/coming to terraform from. Given the existing private dns resources all start `azurerm_private_dns_*` and that the portal also calls it Private resolver users are going to expect/look for "azure private resolver" not "azure dns resolver"

Because of this i concur with everyone else here the name should be `azurerm_private_dns_resolver` to match the portal's chosen name for these resources and match existing resource in terraform. As long as that is its name in the portal (our primary reference for names - not APIs) i don't see us being able to call it something else.

@ms-henglu can we update the name as requested to at the very least get this in for users? we can always add a resource alias later on for a different name if required

File: .github/labeler-pull-request-triage.yml [hunk 1]
```diff
@@ -108,6 +108,9 @@ service/dns:
   - internal/services/dns/**/*
   - internal/services/privatedns/**/*
 
+service/dns-resolver:
+  - internal/services/dnsresolver/**/*
```
Comment (katbyte) [inline .github/labeler-pull-request-triage.yml:112]:
```suggestion
service/private-dns-resolver:
  - internal/services/privatednsresolver/**/*
```

File: .teamcity/components/generated/services.kt [hunk 2]
```diff
@@ -28,6 +28,7 @@ var services = mapOf(
         "costmanagement" to "Cost Management",
         "customproviders" to "Custom Providers",
         "dns" to "DNS",
+        "dnsresolver" to "DNS Resolver",
```
Comment (katbyte) [inline .teamcity/components/generated/services.kt:31]:
```suggestion
        "privatednsresolver" to "Private DNS Resolver",
```

File: .teamcity/components/settings.kt [hunk 3]
```diff
@@ -63,6 +63,9 @@ var serviceTestConfigurationOverrides = mapOf(
         // Data Lake has a low quota
         "datalake" to testConfiguration(parallelism = 2),
 
+        // DNS Resolver is only available in certain locations
+        "dnsresolver" to testConfiguration(locationOverride = LocationConfiguration("eastus", "westus3", "westeurope", true)),
```
Comment (katbyte) [inline .teamcity/components/settings.kt:67]:
```suggestion
        "privatednsresolver" to testConfiguration(locationOverride = LocationConfiguration("eastus", "westus3", "westeurope", true)),
```

File: internal/clients/client.go [hunk 4]
```diff
@@ -165,6 +166,7 @@ type Client struct {
 	DigitalTwins          *digitaltwins.Client
 	Disks                 *disks.Client
 	Dns                   *dns_v2018_05_01.Client
+	DNSResolver           *dnsresolver.Client
```
Comment (katbyte) [inline internal/clients/client.go:169]: and here

File: internal/clients/client.go [hunk 5]
```diff
@@ -286,6 +288,7 @@ func (client *Client) Build(ctx context.Context, o *common.ClientOptions) error
 	client.DigitalTwins = digitaltwins.NewClient(o)
 	client.Disks = disks.NewClient(o)
 	client.Dns = dns.NewClient(o)
+	client.DNSResolver = dnsresolver.NewClient(o)
```
Comment (katbyte) [inline internal/clients/client.go:291]: and here

File: internal/services/dnsresolver/client/client.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package client
```
Comment (katbyte) [inline internal/services/dnsresolver/client/client.go:1]: and the path for the service needs to change

**Review by katbyte — CHANGES_REQUESTED** [review]

**Review by katbyte — APPROVED** [review]
Thanks @ms-henglu - LGTM 🧮

**Comment by DevopsMercenary** [comment]
Don't see any docs in v3.28.0 for azurerm_private_dns_resolver nor in the release notes.

**Comment by ms-henglu** [comment]
Hi @DevopsMercenary ,

I guess it was labeled v3.28.0 by mistake, but these resources have been released in v3.29.0, please check: https://github.com/hashicorp/terraform-provider-azurerm/releases/tag/v3.29.0

**Pull request merged**